From patchwork Fri Sep 21 17:36:17 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10610755
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 3B341157B
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 21 Sep 2018 17:36:28 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 2015D2E3D2
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 21 Sep 2018 17:36:28 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 110462E3D7; Fri, 21 Sep 2018 17:36:28 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id A39C92E3D2
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 21 Sep 2018 17:36:27 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S2390726AbeIUX0T (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 21 Sep 2018 19:26:19 -0400
Received: from mail-io1-f74.google.com ([209.85.166.74]:43344 "EHLO
        mail-io1-f74.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S2390180AbeIUX0T (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 21 Sep 2018 19:26:19 -0400
Received: by mail-io1-f74.google.com with SMTP id w19-v6so23275067ioa.10
        for <kvm@vger.kernel.org>; Fri, 21 Sep 2018 10:36:25 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=4UhaDDzHd1E0S3IM/9TBj6u7QYtVUnSPaOzq3rO1WlI=;
        b=pELi1oWT5AgEGiXLKFt6jjB4QOdiZXqOhAE4U1pmhJrgTuNF1qHBYHzaHkwV2nMhba
         +AaZoZCq7VwVssHzMVX3SyI2eP4ct2vrc/cf/0yC3ntYi8QuziViulkL17BUzYrgxaiH
         O9pAS1e/SWhJjLIwwwZ1Rcql1fqcB+avQ9ZfyhKqr41/cQngZNoTTwcQLXvF6v8HMfa8
         I9k+/fwJGCCU7rPZNlM3f/j5rgBAfCyi6Ri0rqJ6Tdbq8c6zovNjx4fVUKpXa1WPQGpa
         Wr9gX+WTycEc3vD2JDdfXBuTCl1Xf7QNsq2oF/heosx4M/tKSfkqJMvcRmNJzxnGPWwb
         Aaxw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=4UhaDDzHd1E0S3IM/9TBj6u7QYtVUnSPaOzq3rO1WlI=;
        b=tl3bp0U2gdjrZofguFhSyfK773b3+x7hGfoUM/2MWTcAA2TIiufsuPaRV23pp60ElR
         83MLhfpDov3xeXkoJbwbz584+EAwMiXBCZYqPJqion5ulwiY+9MG5CFHvuyvUbH+CVg5
         AC7yHm3SFnhB3Jq7QIF7bfHk6wahwAzO8NM7jEE5rVQtWBvZ7Iduwi7iFomcAo1je2N5
         CIyX3iVU1/B0iqtFfCD4opfhP4fdO2fGQ2APFmIUfidW0pWB6c6Sppkhi/WCtv5woHfo
         9JVz+PJWqh/Q6ROCUYcV8hiY2nwNAfltDzGMRgeCdU+mTmhvhp1fnFsxfovvn6uxmA+c
         4RKQ==
X-Gm-Message-State: APzg51Bt143pfKHrBKe8ns5XgfFpzkIBnwOrgoRRbdILspCt3ux0mxxt
        0EElOSjSnUptbDdkrz93eXaIwP9tf5MrWg4hiM72I2jgGFuYLNIqv9c91m2IJE1m8hh3ra0uwg9
        wX5ZN26XWCxcKpmStWH7SFmoBaVb1z4WGPpS1WEPDpN0tNAi0NpK5Wmd+e3KRelg=
X-Google-Smtp-Source: 
 ANB0Vda1rRmh3wchwRwWzmiPlZb9OhsIyhJc/3waIT3wSYcvqUiQRD039npjU0vQu2g3v2mQBBd5nvALjsNQrg==
X-Received: by 2002:a6b:3c17:: with SMTP id
 k23-v6mr31450379iob.81.1537551384839;
 Fri, 21 Sep 2018 10:36:24 -0700 (PDT)
Date: Fri, 21 Sep 2018 10:36:17 -0700
In-Reply-To: 
 <CALMp9eQzLbHW+WPWcG9=jsR=N=Wo_5bDgwAvJ8W5_HqoouhHdg@mail.gmail.com>
Message-Id: <20180921173617.122426-1-jmattson@google.com>
Mime-Version: 1.0
References: 
 <CALMp9eQzLbHW+WPWcG9=jsR=N=Wo_5bDgwAvJ8W5_HqoouhHdg@mail.gmail.com>
X-Mailer: git-send-email 2.19.0.444.g18242da7ef-goog
Subject: [PATCH v2] KVM: nVMX: Clear reserved bits of #DB exit qualification
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org, Paolo Bonzini <pbonzini@redhat.com>,
        Sean Christopherson <sean.j.christopherson@intel.com>
Cc: Jim Mattson <jmattson@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

According to volume 3 of the SDM, bits 63:15 and 12:4 of the exit
qualification field for debug exceptions are reserved (cleared to
0). However, the SDM is incorrect about bit 16 (corresponding to
DR6.RTM). This bit should be set if a debug exception (#DB) or a
breakpoint exception (#BP) occurred inside an RTM region while
advanced debugging of RTM transactional regions was enabled. Note that
this is the opposite of DR6.RTM, which "indicates (when clear) that a
debug exception (#DB) or breakpoint exception (#BP) occurred inside an
RTM region while advanced debugging of RTM transactional regions was
enabled."

There is still an issue with stale DR6 bits potentially being
misreported for the current debug exception.  DR6 should not have been
modified before vectoring the #DB exception, and the "new DR6 bits"
should be available somewhere, but it was and they aren't.

Fixes: b96fb439774e1 ("KVM: nVMX: fixes to nested virt interrupt injection")
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Sean Christopherson <sean.j.christopherson@intel.com>
---
 arch/x86/include/asm/kvm_host.h | 1 +
 arch/x86/kvm/vmx.c              | 7 +++++--
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/arch/x86/include/asm/kvm_host.h b/arch/x86/include/asm/kvm_host.h
index 09b2e3e2cf1b..1c09a0d1771f 100644
--- a/arch/x86/include/asm/kvm_host.h
+++ b/arch/x86/include/asm/kvm_host.h
@@ -177,6 +177,7 @@ enum {
 
 #define DR6_BD		(1 << 13)
 #define DR6_BS		(1 << 14)
+#define DR6_BT		(1 << 15)
 #define DR6_RTM		(1 << 16)
 #define DR6_FIXED_1	0xfffe0ff0
 #define DR6_INIT	0xffff0ff0
diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 06412ba46aa3..13e41e12945b 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -3297,10 +3297,13 @@ static int nested_vmx_check_exception(struct kvm_vcpu *vcpu, unsigned long *exit
 		}
 	} else {
 		if (vmcs12->exception_bitmap & (1u << nr)) {
-			if (nr == DB_VECTOR)
+			if (nr == DB_VECTOR) {
 				*exit_qual = vcpu->arch.dr6;
-			else
+				*exit_qual &= ~(DR6_FIXED_1 | DR6_BT);
+				*exit_qual ^= DR6_RTM;
+			} else {
 				*exit_qual = 0;
+			}
 			return 1;
 		}
 	}
